File: src/app/view/TextController.java
package app.view;

import app.App;
import javafx.application.Platform;
import javafx.fxml.FXML;
import javafx.scene.control.Alert;
import javafx.scene.control.TextArea;

import java.util.Arrays;
import java.util.Timer;
import java.util.TimerTask;

import static javafx.scene.control.Alert.AlertType.WARNING;

public class TextController {
    @FXML
    private TextArea textInputE;
    @FXML
    private TextArea textInputD;

    private Timer timerE;
    private Timer timerD;

    @FXML
    private synchronized void startEncryptText() {
        clearTextAreaBColor();
        textInputD.setText("");

        if (timerE != null) {
            timerE.cancel();
        }
        timerE = new Timer(true);

        TimerTask timerTaskE = new TimerTask() {
            @Override
            public void run() {
                encryptText();
            }
        };
        timerE.schedule(timerTaskE, 1000);
    }

    private void encryptText() {
        App.getRsa().ifPresentOrElse(
                rsa -> Platform.runLater(() -> {
                    try {
                        textInputD.setText(Arrays.toString(rsa.encryptObj(textInputE.getText())));
                        textInputE.setStyle("text-area-background: rgba(188,255,209,0.5)");
                    } catch (Exception e) {
                        e.printStackTrace();
                        textInputE.setStyle("text-area-background: rgba(255,181,174,0.51)");
                    }
                }),
                () -> Platform.runLater(
                        () -> new Alert(WARNING, "Please choose a RSA key file.").showAndWait()
                )
        );
    }

    @FXML
    private synchronized void startDecryptText() {
        clearTextAreaBColor();
        textInputE.setText("");

        if (timerD != null) {
            timerD.cancel();
        }
        timerD = new Timer(true);

        TimerTask timerTaskD = new TimerTask() {
            @Override
            public void run() {
                decryptText();
            }
        };
        timerD.schedule(timerTaskD, 1000);
    }

    private void decryptText() {
        App.getRsa().ifPresentOrElse(
                rsa -> {
                    int[] integers = Arrays.stream(textInputD.getText().split("\\[|]|\\s+|,"))
                            .filter(s -> s.matches("-?\\d+"))
                            .mapToInt(Integer::valueOf)
                            .toArray();
                    byte[] bytes = new byte[integers.length];
                    for (int i = 0; i < integers.length; i++) {
                        bytes[i] = (byte) integers[i];
                    }

                    Platform.runLater(() -> {
                        try {
                            textInputE.setText(rsa.decryptObj(bytes));
                            textInputD.setStyle("text-area-background: rgba(188,255,209,0.51)");
                        } catch (Exception e) {
                            e.printStackTrace();
                            textInputD.setStyle("text-area-background: rgba(255,181,174,0.5)");
                        }
                    });
                },
                () -> Platform.runLater(
                        () -> new Alert(WARNING, "Please choose a RSA key file.").showAndWait()
                )
        );
    }

    // todo: 将代码中的CSS部分写到css文件里面。
    private void clearTextAreaBColor() {
        textInputE.setStyle("-fx-background-color: #ffffff");
        textInputD.setStyle("-fx-background-color: #ffffff");
    }
}

File: src/app/view/TextController.kt
package app.view

import app.App
import javafx.application.Platform.runLater
import javafx.fxml.FXML
import javafx.scene.control.Alert
import javafx.scene.control.Alert.AlertType.WARNING
import javafx.scene.control.TextArea
import java.util.*

class TextController {
    @FXML
    private lateinit var textInputE: TextArea
    @FXML
    private lateinit var textInputD: TextArea

    private var timerE: Timer? = null
    private var timerD: Timer? = null

    private fun clearTextAreaBColor() {
        textInputE.styleClass.clear()
        textInputD.styleClass.clear()
    }

    @FXML
    private fun startEncryptText() {
        clearTextAreaBColor()
        textInputD.text = ""
        timerE?.cancel()

        timerE = Timer(true)
        timerE!!.schedule(object : TimerTask() {
            override fun run() {
                encryptText()
            }
        }, TIMER_DELAY)
    }

    @FXML
    private fun startDecryptText() {
        clearTextAreaBColor()
        textInputE.text = ""
        timerD?.cancel()

        timerD = Timer(true)
        timerD!!.schedule(object : TimerTask() {
            override fun run() {
                decryptText()
            }
        }, TIMER_DELAY)
    }

    private fun encryptText() {
        App.rsa?.let {
            try {
                val encrypted = it.encryptObj(textInputE.text).contentToString()
                runLater {
                    textInputD.text = encrypted
                    textInputE.setStyleClass(T_SUCCESS)
                }
            } catch (e: Exception) {
                e.printStackTrace()
                runLater {
                    textInputE.setStyleClass(T_WARNING)
                }
            }
        } ?: run {
            runLater {
                Alert(WARNING, "Please choose a RSA key file.").showAndWait()
            }
        }
    }

    private fun decryptText() {
        App.rsa?.let {
            try {
                val bytes = textInputD.text
                        .split("\\[|]|\\s+|,".toRegex())
                        .filter { it.matches("-?\\d+".toRegex()) }
                        .map { it.toByte() }
                        .toByteArray()
                val decrypted = it.decryptObj<String>(bytes)

                runLater {
                    textInputE.text = decrypted
                    textInputD.styleClass.setAll("textarea-success")
                }
            } catch (e: Exception) {
                e.printStackTrace()
                runLater {
                    textInputD.styleClass.setAll("textarea-warning")
                }
            }
        } ?: run {
            Alert(WARNING, "Please choose a RSA key file.").showAndWait()
        }
    }

    companion object {
        private const val TIMER_DELAY = 1000L
    }
}
